transports/quic/src/hole_punching.rs

```diff
@@ -0,0 +1,95 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{channel::oneshot, prelude::*};
+use libp2p_identity::PeerId;
+use rand::{distributions, Rng};
+
+#[cfg(feature = "async-std")]
+use async_std::{future::timeout, task::sleep};
+#[cfg(feature = "tokio")]
+use tokio::time::{sleep, timeout};
+
+use crate::{
+    endpoint::{self, ToEndpoint},
+    Connecting, Connection, Error,
+};
+
+pub(crate) type HolePunchMap =
+    Arc<Mutex<HashMap<(SocketAddr, PeerId), oneshot::Sender<(PeerId, Connection)>>>>;
+
+/// An upgrading inbound QUIC connection that is either
+/// - a normal inbound connection or
+/// - an inbound connection corresponding to an in-progress outbound hole punching connection.
+pub(crate) struct MaybeHolePunchedConnection {
+    hole_punch_map: HolePunchMap,
+    addr: SocketAddr,
+    upgrade: Connecting,
+}
+
+impl MaybeHolePunchedConnection {
+    pub(crate) fn new(hole_punch_map: HolePunchMap, addr: SocketAddr, upgrade: Connecting) -> Self {
+        Self {
+            hole_punch_map,
+            addr,
+            upgrade,
+        }
+    }
+}
+
+impl Future for MaybeHolePunchedConnection {
+    type Output = Result<(PeerId, Connection), Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let (peer_id, connection) = futures::ready!(self.upgrade.poll_unpin(cx))?;
+        let addr = self.addr;
+        let mut hole_punch_map = self.hole_punch_map.lock().unwrap();
+        if let Some(sender) = hole_punch_map.remove(&(addr, peer_id)) {
+            if let Err(connection) = sender.send((peer_id, connection)) {
+                Poll::Ready(Ok(connection))
+            } else {
+                Poll::Ready(Err(Error::HandshakeTimedOut))
+            }
+        } else {
+            Poll::Ready(Ok((peer_id, connection)))
+        }
+    }
+}
```

Comment:
Couldn't this also just be an `async fn`? Or perhaps just a `.map` on `Connecting`?